src/core/providers/copc/qgscopcprovider.cpp

```diff
@@ -0,0 +1,240 @@
+{
+  return mIndex->metadataClassStatistic( attribute, value, statistic );
+}
+
+void QgsCopcProvider::loadIndex( )
+{
+  if ( mIndex->isValid() )
+    return;
+
+  mIndex->load( dataSourceUri() );
+
+  // TODO: Read metadata from LAZ into mOriginalMetadata
+  // NOTE: This is not done in EPT provider either
+}
+
+QVariantMap QgsCopcProvider::originalMetadata() const
+{
+  return mOriginalMetadata;
+}
+
+void QgsCopcProvider::generateIndex()
+{
+  //no-op, index is always generated
+}
+
+QVariant QgsCopcProvider::metadataStatistic( const QString &attribute, QgsStatisticalSummary::Statistic statistic ) const
+{
+  return mIndex->metadataStatistic( attribute, statistic );
+}
+
+QgsCopcProviderMetadata::QgsCopcProviderMetadata():
+  QgsProviderMetadata( PROVIDER_KEY, PROVIDER_DESCRIPTION )
+{
+}
+
+QgsCopcProvider *QgsCopcProviderMetadata::createProvider( const QString &uri, const QgsDataProvider::ProviderOptions &options, QgsDataProvider::ReadFlags flags )
+{
+  return new QgsCopcProvider( uri, options, flags );
+}
+
+QList<QgsProviderSublayerDetails> QgsCopcProviderMetadata::querySublayers( const QString &uri, Qgis::SublayerQueryFlags, QgsFeedback * ) const
+{
+  const QVariantMap parts = decodeUri( uri );
+  if ( parts.value( QStringLiteral( "path" ) ).toString().endsWith( ".copc.laz", Qt::CaseSensitivity::CaseInsensitive ) )
+  {
+    QgsProviderSublayerDetails details;
+    details.setUri( uri );
+    details.setProviderKey( QStringLiteral( "copc" ) );
+    details.setType( QgsMapLayerType::PointCloudLayer );
+    details.setName( QgsProviderUtils::suggestLayerNameFromFilePath( uri ) );
+    return {details};
+  }
+  else
+  {
+    return {};
+  }
+}
+
+int QgsCopcProviderMetadata::priorityForUri( const QString &uri ) const
+{
+  const QVariantMap parts = decodeUri( uri );
+  const QFileInfo fi( parts.value( QStringLiteral( "path" ) ).toString() );
+  if ( parts.value( QStringLiteral( "path" ) ).toString().endsWith( ".copc.laz", Qt::CaseSensitivity::CaseInsensitive ) )
+    return 101;
+
+  return 0;
+}
+
+QList<QgsMapLayerType> QgsCopcProviderMetadata::validLayerTypesForUri( const QString &uri ) const
+{
+  const QVariantMap parts = decodeUri( uri );
+  const QFileInfo fi( parts.value( QStringLiteral( "path" ) ).toString() );
+  if ( parts.value( QStringLiteral( "path" ) ).toString().endsWith( ".copc.laz", Qt::CaseSensitivity::CaseInsensitive ) )
+    return QList< QgsMapLayerType>() << QgsMapLayerType::PointCloudLayer;
+
+  return QList< QgsMapLayerType>();
+}
+
+bool QgsCopcProviderMetadata::uriIsBlocklisted( const QString &uri ) const
+{
+  const QVariantMap parts = decodeUri( uri );
+  if ( !parts.contains( QStringLiteral( "path" ) ) )
+    return false;
+
+  const QFileInfo fi( parts.value( QStringLiteral( "path" ) ).toString() );
+
+  // internal details only
+  if ( parts.value( QStringLiteral( "path" ) ).toString().endsWith( ".copc.laz", Qt::CaseSensitivity::CaseInsensitive ) )
```

Comment:
This logic is wrong, here you are saying that a ".copc.laz" file should never be exposed to users.

This should be handled in QgsPdalProviderMetadata instead -- in QgsPdalProviderMetadata::priorityForUri/validLayerTypesForUri/querySubLayers you should adapt the existing tests so that .copc.laz files aren't accepted. Right now QgsPdalProviderMetadata just checks for a .laz extension, so it's greedily grabbing the .copc.laz uris.